### ISSUE TITLE
`Tuple index out of range` doesn't take into account `len` check when accessing `Union` of `Tuples` of varying lengths

### ISSUE THREAD

**Issue by posita** [python/mypy#13941]
Where a type is a `Union` of `Tuples` of varying lengths, Mypy remains blissfully unaware of any `len` validation before accessing elements beyond the minimum length.

``` python
# test_case.py
from typing import Optional, Tuple, Union

def get_third_thing_works_correctly(thing: Tuple[int, int, Optional[int]]):
    return thing[2]

def get_third_thing_errors_correctly(thing: Tuple[int, int]):
    return thing[2]  # <-- correctly results in "Tuple index out of range [misc]"

def get_third_thing_false_positive(
    thing: Union[Tuple[int, int], Tuple[int, int, Optional[int]]]
):
    if len(thing) >= 3:
        return thing[2]  # <-- false positive
    else:
        return None
```

Results:

``` sh
% python --version
Python 3.10.6
% mypy --version
mypy 0.982 (compiled: yes)
% mypy --config=/dev/null test_case.py
/dev/null: No [mypy] section in config file
test_case.py:8: error: Tuple index out of range
test_case.py:14: error: Tuple index out of range
Found 2 errors in 1 file (checked 1 source file)
```

**Comment by AlexWaygood** [comment]
This is arguably a duplicate of #1178, though it's a slightly different sympton of the same problem

**Comment by JelleZijlstra** [comment]
It's the same thing: a feature request that we use `len()` for type narrowing. Closing as a duplicate.
